README

Copyright Richard Vaughan and contributors 1998-2008.  
Part of the Player Project (http://playerstage.sourceforge.net)


License
-------
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.
 
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

A copy of the license is included with the sourcecode in the file
'COPYING". Copying and redistribution is permitted only under the
terms of the license.


Introduction
------------
Stage is a robot simulator. It provides a virtual world populated by
mobile robots and sensors, along with various objects for the robots
to sense and manipulate.

There are three ways to use Stage: 

1. The "stage" program: a standalone robot simulation program
that loads your robot control program from a library that you provide.

2. The Stage plugin for Player (libstageplugin) - provides a
population of virtual robots for the popular Player networked robot
interface system.

3. Write your own simulator: the "libstage" C++ library makes it
easy to create, run and customize a Stage simulation from inside your
own programs.


Models
------
Stage provides several sensor and actuator models, including sonar
or infrared rangers, scanning laser rangefinder, color-blob tracking,
fiducial tracking, bumpers, grippers and mobile robot bases with
odometric or global localization.


Design
------
Stage was designed with multi-agent systems in mind, so it provides
fairly simple, computationally cheap models of lots of devices rather
than attempting to emulate any device with great fidelity. This design
is intended to be useful compromise between conventional high-fidelity
robot simulations, the minimal simulations described by Jakobi [4], and
the grid-world simulations common in artificial life research [5]. We
intend Stage to be just realistic enough to enable users to move
controllers between Stage robots and real robots, while still being
fast enough to simulate large populations. We also intend Stage to be
comprehensible to undergraduate students, yet sophisticated enough for
professional reseachers.
  
Player also contains several useful 'virtual devices'; including
some sensor pre-processing or sensor-integration algorithms that help
you to rapidly build powerful robot controllers. These are easy to use
with Stage.


Authors
-------
Stage was written by:

- Richard Vaughan (vaughan@sfu.ca)
- Brian Gerkey (gerkey@willowgarage.com)
- Reed Hedges (hedges@mobilerobots.com)
- Andrew Howard (andrew.b.howard@jpl.nasa.gov)
- Toby Collett (tcollett+player@plan9.net.nz)
- Pooya Karimian (pooyak@sfu.ca)
- Jeremy Asher (jra11@sfu.ca)
- Alex Couture-Beil (asc17@sfu.ca)


Many patches and bug reports have been contributed by users around the
world.

Stage is part of the Player Project (http://playerstage.sourceforge.net),
a community effort to develop Free Software tools for robotics research.


Citations
---------
If you use Stage in your work, we'd appreciate a citation. At the time 
of writing, the most suitable reference is:

Brian Gerkey, Richard T. Vaughan and Andrew Howard. "The Player/Stage 
Project: Tools for Multi-Robot and Distributed Sensor Systems" 
Proceedings of the 11th International Conference on Advanced Robotics,
pages 317-323, Coimbra, Portugal, June 2003 (ICAR'03)  
http://www.isr.uc.pt/icar03/ .  

[gzipped postscript](http://robotics.stanford.edu/~gerkey/research/final_papers/icar03-player.ps.gz), 
[pdf](http://robotics.stanford.edu/~gerkey/research/final_papers/icar03-player.pdf)

Please help us keep track of what's being used out there by correctly 
naming the Player/Stage components you use. Player used on its own is 
called "Player". Player and Stage used together are referred to as 
"the Player/Stage system" or just "Player/Stage". When libstage is 
used without Player, it's just called "Stage". When Player is used 
with its 3D ODE-based simulation backend, Gazebo, it's called 
Player/Gazebo. Gazebo without Player is just "Gazebo". All this 
software is part of the "Player Project".


References
----------



Support
-------
Funding for Stage has been provided in part by:

- DARPA (USA)
- NASA (USA)
- NSERC (Canada)
- NSF (USA)
- Simon Fraser University (Canada)


Names
-----
The names "Player" and "Stage" were inspired by the lines:

    "All the world's a stage,  
     And all the men and women merely players"  

from "As You Like It" by William Shakespeare.


References
----------
[4] Nick Jakobi (1997) "Evolutionary Robotics and the Radical Envelope
of Noise Hypothesis", Adaptive Behavior Volume 6, Issue 2. pp.325 -
368.
 
[5] Stuart Wilson (1985) "Knowledge Growth in an Artificial Animal",
Proceedings of the First International Conference on Genetic Agorithms
and Their Applications.  Hillsdale, New Jersey. pp.16-23.

ChangeLog
2009-01-16: thjc
	* 2.1.1 release
		Fixes for gcc 4.3 compilation
		fix for close of null file pointer
		patch 2187145 - Added wheelbase option to position "car" model
		patch 2234440 - makes blobfinder ignore colour alpha value
		patch 2230769 - possible infinite loop in stg_model_update_pose
		patch 2150737 - bumper detects collisions incorrectly
		patch 2475933 - Allow graphics2d with gui disabled
		patch 2124895, fix for subscription to simulation device
		patch 2286925 DIO interface and indicator model

2008-06-14: thjc
	* 2.1.0 release
		Player 2.1 compatibility release

2008-04-16 18:47  thjc
	* 2.1rc2 release

2007-12-13 13:07  gerkey

	* stage.pc.in (1.8), src/Makefile.am (1.144): fixed up lib version
	  numbers and changed headers to go into a versioned include
	  directory

2007-12-13 12:38  gerkey

	* worlds/wifi/: .cvsignore (1.1), Makefile.am (1.1): added build
	  files

2007-12-13 12:31  gerkey

	* configure.ac (1.17), worlds/Makefile.am (1.20): incremented
	  version, updated example installation

2007-12-11 15:23  gerkey

	* src/: model_ptz.c (1.5), p_driver.cc (1.43): fixed ptz and
	  graphics2d behavior when gui_disable is on

2007-12-11 11:42  gerkey

	* src/: gui_menus.c (1.73), model.c (1.156): backported gui_disable
	  fix from HEAD

2007-12-10 11:41  gerkey

	* worlds/: wavefront-remote.cfg (1.1), wavefront.cfg (1.13): added
	  more examples

2007-12-05 13:51  gerkey

	* src/p_wifi.cc (1.5): tweak to wifi model

2007-11-26 13:50  gerkey

	* src/p_wifi.cc (1.4): started adding publishing support to wifi
	  model

2007-11-01 18:11  gerkey

	* worlds/simple.cfg (1.34): added map export simple.cfg

2007-11-01 17:06  gerkey

	* src/p_simulation.cc (1.23): added normalization to simulation
	  interface angle

2007-11-01 15:15  gerkey

	* worlds/simple.cfg (1.33), src/p_blobfinder.cc (1.11),
	  src/p_bumper.cc (1.4), src/p_fiducial.cc (1.11),
	  src/p_graphics2d.cc (1.7), src/p_laser.cc (1.21), src/p_map.cc
	  (1.14), src/p_sonar.cc (1.11): applied Toby's patch to replace
	  fixed-size arrays

2007-09-11 14:30  rtv

	* src/gui.h (1.37), src/model_wifi.cc (1.2), src/raytrace.c (1.18),
	  src/stage.h (1.193), src/stage_internal.h (1.60),
	  worlds/wifi/commando.cfg (1.1), worlds/wifi/commando.world (1.1),
	  worlds/wifi/hosp_wifi.cfg (1.1), worlds/wifi/hosp_wifi.world
	  (1.1), worlds/wifi/hosp_wifi_5.cfg (1.1),
	  worlds/wifi/hosp_wifi_5.world (1.1), worlds/wifi/map.inc (1.1),
	  worlds/wifi/wifi.cfg (1.1), worlds/wifi/wifi.world (1.1),
	  worlds/wifi/wifi_itu.cfg (1.1), worlds/wifi/wifi_itu.world (1.1),
	  worlds/wifi/wifi_logdistance.cfg (1.1),
	  worlds/wifi/wifi_logdistance.world (1.1),
	  worlds/wifi/wifi_ray.cfg (1.1), worlds/wifi/wifi_ray.world (1.1),
	  worlds/wifi/wifi_simple.cfg (1.1), worlds/wifi/wifi_simple.world
	  (1.1): added Wifi model, more or less untested, from Michael
	  Schroeder. Comments welcome

2007-09-11 13:39  rtv

	* src/p_driver.cc (1.42), worlds/pioneer.inc (1.29): applied player
	  driver timing fix from Alex Brooks

2007-09-10 19:13  rtv

	* .cvsignore (1.9): tweaked cvsignore

2007-09-10 18:24  rtv

	* worlds/: map.inc (1.13), pioneer.inc (1.28), sick.inc (1.3),
	  playerswarm/Makefile (1.1), playerswarm/args.h (1.1),
	  playerswarm/many.sh (1.1), playerswarm/multiple.cc (1.1),
	  playerswarm/pss.cfg (1.1), playerswarm/pss.world (1.1),
	  playerswarm/pss3d.cfg (1.1), playerswarm/pss3d.world (1.1),
	  playerswarm/pss_large.cfg (1.1), playerswarm/pss_large.world
	  (1.1), playerswarm/single.cc (1.1): added 100-robot swarm demo.
	  Useful for testing

2007-09-10 18:14  rtv

	* INSTALL (1.8), bootstrap (1.11), src/gui_menus.c (1.72),
	  src/model_fiducial.c (1.51), src/p_driver.cc (1.41),
	  src/p_driver.h (1.25), src/p_gripper.cc (1.10): tweaks for better
	  scalability and some bugfixes

2007-09-05 12:01  gerkey

	* src/: model_ptz.c (1.4), p_ptz.cc (1.4), stage.h (1.192): applied
	  patch to support velocity control of ptz

2007-08-23 12:58  gerkey

	* src/: p_audio.cc (1.5), p_blobfinder.cc (1.10), p_bumper.cc
	  (1.3), p_driver.cc (1.40), p_driver.h (1.24), p_fiducial.cc
	  (1.10), p_graphics2d.cc (1.6), p_gripper.cc (1.9), p_laser.cc
	  (1.20), p_localize.cc (1.3), p_map.cc (1.13), p_position.cc
	  (1.18), p_power.cc (1.4), p_ptz.cc (1.3), p_simulation.cc (1.22),
	  p_sonar.cc (1.10), p_speech.cc (1.2), p_wifi.cc (1.3): applied
	  Toby's autopointer patches

2007-08-21 16:28  gerkey

	* worlds/: mbicp.cfg (1.1), mbicp.world (1.1), p3at.inc (1.1),
	  p3dx.inc (1.1), pioneerBase.inc (1.1), sick.inc (1.2),
	  bitmaps/mbicp.png (1.1): added mbicp examples

2007-08-21 13:28  gerkey

	* src/p_position.cc (1.17): fixed return values from processmessage

2007-07-09 10:18  gerkey

	* src/: p_blobfinder.cc (1.9), p_bumper.cc (1.2), p_fiducial.cc
	  (1.9), p_laser.cc (1.19), p_position.cc (1.16), p_sonar.cc (1.9),
	  p_wifi.cc (1.2): applied patch 1749747 to transition to 3d pose
	  structures

2007-05-14 17:08  gerkey

	* src/model_ptz.c (1.3): added fix for zero panspeed

2007-04-30 12:51  gerkey

	* worlds/nd.cfg (1.1): added nd example

2007-04-17 14:08  gerkey

	* worlds/Makefile.am (1.19): added roomba.inc to installation

2007-03-26 12:54  gerkey

	* worlds/: everything.cfg (1.34), everything.world (1.86),
	  roomba.inc (1.1): added roomba to everything.world

2006-12-01 16:41  rtv

	* src/p_simulation.cc (1.21): fixed incompatibity with Player HEAD

2006-11-20 17:48  rtv

	* src/model_laser.c (1.90): fixed bug in handling laser_bright
	  setting. Thanks to mzuluaga.

2006-11-20 17:44  rtv

	* src/model_load.c (1.11): applied a patch to fix worldfile angle
	  loading.  Thanks to reddish.

2006-11-20 17:30  rtv

	* bootstrap (1.10), src/gui.c (1.110), src/matrix.c (1.23),
	  src/model.c (1.155), src/model_fiducial.c (1.50),
	  src/model_load.c (1.10), src/p_simulation.cc (1.20), src/rtk.c
	  (1.10), src/rtk_canvas.c (1.22), src/rtk_fig.c (1.22),
	  src/stage.c (1.90), src/world.c (1.64): applied patches from user

src/gui.h

// GUI constants

#define STG_LAYER_BACKGROUND 10
#define STG_LAYER_BODY 30
#define STG_LAYER_GRID 20
#define STG_LAYER_USER 99
#define STG_LAYER_GEOM 80
#define STG_LAYER_MATRIX 40
#define STG_LAYER_MATRIX_TREE 39
#define STG_LAYER_DEBUG 98

#define STG_LAYER_BLOBGEOM 51
#define STG_LAYER_BLOBCONFIG 52
#define STG_LAYER_BLOBDATA 53

#define STG_LAYER_PTZGEOM 54
#define STG_LAYER_PTZONFIG 56
#define STG_LAYER_PTZDATA 57

#define STG_LAYER_NEIGHBORGEOM 61
#define STG_LAYER_NEIGHBORCONFIG 62
#define STG_LAYER_NEIGHBORDATA 63

#define STG_LAYER_ENERGYGEOM 71
#define STG_LAYER_ENERGYCONFIG 72
#define STG_LAYER_ENERGYDATA 73

#define STG_GRID_MAJOR_COLOR "gray85"
#define STG_GRID_MINOR_COLOR "gray95"
#define STG_GRID_AXIS_COLOR "gray40"
#define STG_BACKGROUND_COLOR "ivory"
#define STG_BOUNDINGBOX_COLOR "magenta"
#define STG_MATRIX_COLOR "dark green"

  // model color defaults
#define STG_FIDUCIAL_COLOR "lime green"
#define STG_FIDUCIAL_CFG_COLOR "green"

#define STG_ENERGY_COLOR "purple"
#define STG_ENERGY_CFG_COLOR "magenta"
   
#define STG_DEBUG_COLOR "red"
#define STG_BLOB_CFG_COLOR "gray75"

#define STG_LAYER_RANGERGEOM 41
#define STG_LAYER_RANGERCONFIG 42
#define STG_LAYER_RANGERDATA 43

#define STG_LAYER_BUMPERGEOM 60
#define STG_LAYER_BUMPERCONFIG 61
#define STG_LAYER_BUMPERDATA 62

#define STG_LAYER_WIFICONFIG 64
#define STG_LAYER_WIFIDATA 65

#define STG_RANGER_COLOR "gray75" 
#define STG_RANGER_GEOM_COLOR "orange"
#define STG_RANGER_CONFIG_COLOR "gray90"

#define STG_LAYER_LASERGEOM 32
#define STG_LAYER_LASERDATA 31
#define STG_LAYER_LASERCONFIG 32

#define STG_LAYER_POSITIONGEOM 45
#define STG_LAYER_POSITIONDATA 47
#define STG_LAYER_POSITIONCONFIG 46

#define STG_LASER_COLOR "steel blue"
#define STG_LASER_FILL_COLOR "powder blue"
#define STG_LASER_GEOM_COLOR "blue"
#define STG_LASER_CFG_COLOR "light steel blue"
#define STG_LASER_BRIGHT_COLOR "blue"

#define STG_WIFI_CFG_COLOR "royal blue"
#define STG_WIFI_CFG_COLOR_GREEN "DarkGreen"


#define STG_LAYER_GRIPPERGEOM 85
#define STG_LAYER_GRIPPERDATA 87
#define STG_LAYER_GRIPPERCONFIG 86

#define STG_GRIPPER_COLOR "steel blue"
#define STG_GRIPPER_FILL_COLOR "light blue"
#define STG_GRIPPER_GEOM_COLOR "blue"
#define STG_GRIPPER_CFG_COLOR "light steel blue"
#define STG_GRIPPER_BRIGHT_COLOR "blue"

  // GUI colors   
#define STG_GRID_MAJOR_COLOR "gray85"
#define STG_GRID_MINOR_COLOR "gray95"
#define STG_MATRIX_COLOR "dark green"
#define STG_BACKGROUND_COLOR "ivory"
  
  // model color defaults
#define STG_GENERIC_COLOR "black"
#define STG_POSITION_COLOR "red"
#define STG_WALL_COLOR "dark blue"
#define STG_SONAR_COLOR "gray70"

#define STG_LAYER_SPEECHDATA 88
#define STG_SPEECH_COLOR "navy"

#define STG_LAYER_AUDIODATA 84
#define STG_AUDIO_COLOR "orange3"
#define STG_AUDIO_BUBBLE_COLOR "orange4"
